sys/drv/rtl8139.c

```diff
@@ -32,18 +39,63 @@ static int rtl_reset(rtl8139_state_t *state) {
   return 1;
 }
 
-static int rtl8139_attach(device_t *dev) {
+static intr_filter_t rtl8139_intr(void *data) {
+  rtl8139_state_t *state = data;
+  // TODO: add enqueue routines
+  bus_write_2(state->regs, RL_ISR, RL_ISR_RX_OK);
+  return IF_STRAY;
+}
+
+static int set_mem_contig_address(rtl8139_state_t *state) {
+  paddr_t page2; /* address of second allocated page */
+
+  /* set and check if any address is assigned */
+  if (!pmap_kextract(state->rx_buf, &state->paddr))
+    return -1;
+
+  /* check if memory is contiguous */
+  if (!pmap_kextract(state->rx_buf + PAGESIZE, &page2))
+    return -1;
+
+  if (state->paddr + PAGESIZE != page2)
+    return -1;
+
+  return 0;
+}
+
+int rtl8139_attach(device_t *dev) {
   rtl8139_state_t *state = dev->state;
 
+  pci_enable_busmaster(dev);
   state->regs = device_take_memory(dev, 1, RF_ACTIVE);
-  if (state->regs == NULL)
+
+  state->irq_res = device_take_irq(dev, 0, RF_ACTIVE);
+  if (!state->regs || !state->irq_res) {
+    klog("rtl8139 failed to init resources");
+    return ENXIO;
+  }
+  bus_intr_setup(dev, state->irq_res, rtl8139_intr, NULL, state, "RTL8139");
+
+  // TODO: mark this memory as PMAP_NOCACHE
+  state->rx_buf = (vaddr_t)kmem_alloc(RX_BUF_SIZE, M_ZERO);
+  if (set_mem_contig_address(state)) {
+    klog("failed to alloc contig memory");
     return ENXIO;
```

Comment:
Here you always allocate memory with success - it's ok, but in case of non contiguous allocation you get error - it's still ok - but you have memory leak.
Your log message is confusing here - "failed to alloc ... memory" and you return `ENXIO`.
- `ENXIO` - Device not configured
- `ENOMEM` - Cannot allocate memory

I would be grateful if you could make that part of your code more obvious for other people.

Reply:
@cahirwpz has improved this part of code with kmem_alloc_contig - thank you
@pj1031999 done - thanks